src/components/KeywordHighlighterComponent.js

```diff
@@ -0,0 +1,255 @@
+/**
+ * Used for turning the highlight on/off.
+ */
+const TOPIC_HIGHLIGHT_ACTIVE_CLASS = "topic-highlighted";
+
+/**
+ * Listens for 'onTextLayerRendered' events and highlights the text using some keywords.
+ * Sentences that should be highlighted are marked by 'span' html elements with class
+ * 'topic-highlighted-text'.
+ *
+ * Class 'topic-highlighted' actually contains css that highlights them.
+ * 2 classes are used so highlighting can be turned off.
+ *
+ * @property {String[]} keywords a list of keywords that are going to be used for
+ *  highlighting
+ * @property {bool} on flag which is going to be used to check whether to highlight
+ *  the appropriate sentences
+ */
+class KeywordHighlighterComponent {
+  constructor() {
+    this.keywords = [];
+    this.on = false;
+
+    this.#registerEvents();
+  }
+
+  /**
+   * Sets the keywords that the highlighting is going to use.
+   *
+   * @param {String[]} keywords
+   */
+  setKeywords(keywords) {
+    this.keywords = keywords;
+  }
+
+  /**
+   * Shows the highlited sentences, text layers processed next will have their
+   * highlited sentences turned on by default.
+   */
+  turnOn() {
+    this.on = true;
+    let highlightedTexts = document.querySelectorAll(
+      `.${TOPIC_HIGHLIGHTED_TEXT_CLASS}`
+    );
+    highlightedTexts.forEach((highlightedText) => {
+      highlightedText.classList.add(TOPIC_HIGHLIGHT_ACTIVE_CLASS);
+    });
+  }
+
+  /**
+   * Hides the highlited sentences, text layers processed next will have their
+   * highlited sentences turned off by default.
+   */
+  turnOff() {
+    this.on = false;
+    let highlightedTexts = document.querySelectorAll(
+      `.${TOPIC_HIGHLIGHTED_TEXT_CLASS}`
+    );
+    highlightedTexts.forEach((highlightedText) => {
+      highlightedText.classList.remove(TOPIC_HIGHLIGHT_ACTIVE_CLASS);
+    });
+  }
+
+  /**
+   * Registers for 'onTextLayerRendered' events to process the rendered text layer
+   * and inject HTML to highlight the sentences.
+   */
+  #registerEvents() {
+    EventHandlerService.subscribe(
+      PDFLEvents.onTextLayerRendered,
+      this.#onTextLayerRendered.bind(this)
+    );
+  }
+
+  /**
+   * Takes a text layer generated by pdf.js library, processes it by injecting
+   * HTML spans to sentences that need to be highlighted.
+   *
+   * Text layer consists of a list of spans called 'presentations'. Presentations
+   * are span elements containing text. There is no rule for how long the text is
+   * going to be in these presentations, or where the end of the sentence is.
+   *
+   * This function goes through all the child 'presentations' and if one contains
+   * a keyword, it will delegate work to other helper functions for injecting HTML
+   * span elements around text that needs to be highlighted.
+   *
+   * @param {HTMLElement} textLayer
+   */
+  #onTextLayerRendered(textLayer) {
+    for (let currentPresentation of textLayer.children) {
+      if (currentPresentation.tagName == "SPAN") {
```

Comment:
You can filter out this collection so you can lose this if and code will be imo more readable, something like
const spans = textLayer.children.filter(currentPresentation => currentPresentation.tagName == "SPAN")

Reply:
That won't work, filter creates a copy of objects. I need the real deal because I mutate them later.

Reply:
No it does not, it only creates new array, but objects in it are not cloned.

Reply:
Lol you know why it "does not work" - because filter expects to be called on array and what you have here is HTMLCollection, it will work if you turn that collection to array like:
const spans = [...textLayer.children].filter(currentPresentation => currentPresentation.tagName == "SPAN");

Like I said in other comment try to lose few indents here. This method looks scary like this and does not go by principles of clean code for "Blocks and Indenting" which says:
"This implies that the blocks within if statements, else statements, while statements, and so on should be one line long. Probably that line should be a function call. Not only does this keep the enclosing function small, but also adds documentary value because the function called within the block can have a nicely descriptive name.
This also implies that functions should not be large enough to hold nested structures. Therefore, the indent level of a function should not be greater than one or two. This, of course, makes the functions easy to read and understand."